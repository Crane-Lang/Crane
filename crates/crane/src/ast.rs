use smol_str::SmolStr;
use thin_vec::ThinVec;

/// An identifier.
#[derive(Debug, Clone)]
pub struct Ident(pub SmolStr);

impl std::fmt::Display for Ident {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}

/// The kind of a Crane expression.
#[derive(Debug, Clone)]
pub enum ExprKind {
    /// A reference to a variable.
    Variable { name: Ident },

    /// A function call.
    Call {
        fun: Box<Expr>,
        args: ThinVec<Box<Expr>>,
    },
}

/// A Crane expression.
#[derive(Debug, Clone)]
pub struct Expr {
    pub kind: ExprKind,
}

/// The kind of a Crane statement.
#[derive(Debug, Clone)]
pub enum StmtKind {
    /// A function declaration (`fn`).
    Fn(Box<Fn>),

    /// An expression.
    Expr(Expr),
}

/// A function definition.
#[derive(Debug, Clone)]
pub struct Fn {
    pub name: SmolStr,
    pub body: ThinVec<Stmt>,
}

/// A Crane statement.
#[derive(Debug, Clone)]
pub struct Stmt {
    pub kind: StmtKind,
}
