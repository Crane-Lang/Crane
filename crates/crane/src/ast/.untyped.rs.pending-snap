{"run_id":"1687112702-689890000","line":225,"new":{"module_name":"crane__ast__untyped__tests","snapshot_name":"ast_node_sizes-4","metadata":{"source":"crates/crane/src/ast/untyped.rs","assertion_line":225,"expression":"size_of::<Item>().to_string()"},"snapshot":"64"},"old":{"module_name":"crane__ast__untyped__tests","metadata":{},"snapshot":"56"}}
{"run_id":"1687112777-693774000","line":222,"new":null,"old":null}
{"run_id":"1687112777-693774000","line":223,"new":null,"old":null}
{"run_id":"1687112777-693774000","line":224,"new":null,"old":null}
{"run_id":"1687112777-693774000","line":225,"new":{"module_name":"crane__ast__untyped__tests","snapshot_name":"ast_node_sizes-4","metadata":{"source":"crates/crane/src/ast/untyped.rs","assertion_line":225,"expression":"size_of::<Item>().to_string()"},"snapshot":"64"},"old":{"module_name":"crane__ast__untyped__tests","metadata":{},"snapshot":"56"}}
{"run_id":"1687112794-630697000","line":222,"new":null,"old":null}
{"run_id":"1687112794-630697000","line":223,"new":null,"old":null}
{"run_id":"1687112794-630697000","line":224,"new":null,"old":null}
{"run_id":"1687112794-630697000","line":225,"new":null,"old":null}
{"run_id":"1687112794-630697000","line":226,"new":null,"old":null}
{"run_id":"1687112794-630697000","line":227,"new":null,"old":null}
{"run_id":"1687112794-630697000","line":228,"new":null,"old":null}
