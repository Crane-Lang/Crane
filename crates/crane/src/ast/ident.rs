use smol_str::SmolStr;

// TODO: Add span to `Ident`.
/// An identifier.
#[derive(Debug, PartialEq, Eq, Hash, Clone)]
pub struct Ident(pub SmolStr);

impl std::fmt::Display for Ident {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}
