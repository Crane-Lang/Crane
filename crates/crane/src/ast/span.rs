use serde::{Deserialize, Serialize};

/// A dummy [`Span`].
///
/// All of the positions are set to `0`.
pub const DUMMY_SPAN: Span = Span { start: 0, end: 0 };

/// A span.
#[derive(PartialEq, Eq, Hash, Clone, Copy, Serialize, Deserialize)]
pub struct Span {
    /// The start offset of the span (inclusive).
    pub start: usize,

    /// The end offset of the span (exclusive).
    pub end: usize,
}

impl Span {
    pub fn new(start: usize, end: usize) -> Self {
        Self { start, end }
    }

    /// Returns a new [`Span`] that encloses `self and `end`.
    pub fn to(self, end: Span) -> Self {
        Self::new(
            std::cmp::min(self.start, end.start),
            std::cmp::max(self.end, end.end),
        )
    }
}

impl std::fmt::Debug for Span {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{:?}..{:?}", self.start, self.end)
    }
}

impl std::fmt::Display for Span {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}..{}", self.start, self.end)
    }
}

impl From<std::ops::Range<usize>> for Span {
    fn from(value: std::ops::Range<usize>) -> Self {
        Self {
            start: value.start,
            end: value.end,
        }
    }
}

impl From<Span> for std::ops::Range<usize> {
    fn from(value: Span) -> Self {
        value.start..value.end
    }
}

impl ariadne::Span for Span {
    type SourceId = ();

    fn source(&self) -> &Self::SourceId {
        &()
    }

    fn start(&self) -> usize {
        self.start
    }

    fn end(&self) -> usize {
        self.end
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn span_to_works_forwards() {
        let start = Span::new(0, 3);
        let end = Span::new(4, 6);

        assert_eq!(start.to(end), Span::new(0, 6))
    }

    #[test]
    fn span_to_works_backwards() {
        let start = Span::new(3, 7);
        let end = Span::new(0, 2);

        assert_eq!(start.to(end), Span::new(0, 7))
    }
}
