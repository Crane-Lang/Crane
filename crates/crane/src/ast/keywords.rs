use smol_str::SmolStr;

use crate::ast::{Ident, DUMMY_SPAN};

pub const FN: Ident = Ident {
    name: SmolStr::new_inline("fn"),
    span: DUMMY_SPAN,
};

pub const LET: Ident = Ident {
    name: SmolStr::new_inline("let"),
    span: DUMMY_SPAN,
};

pub const MOD: Ident = Ident {
    name: SmolStr::new_inline("mod"),
    span: DUMMY_SPAN,
};

pub const PUB: Ident = Ident {
    name: SmolStr::new_inline("pub"),
    span: DUMMY_SPAN,
};

pub const STRUCT: Ident = Ident {
    name: SmolStr::new_inline("struct"),
    span: DUMMY_SPAN,
};

pub const UNION: Ident = Ident {
    name: SmolStr::new_inline("union"),
    span: DUMMY_SPAN,
};
