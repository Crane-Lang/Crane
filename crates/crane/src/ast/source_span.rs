use crate::ast::Span;

/// A [`Span`] with the source from where it originated.
pub struct SourceSpan {
    pub source: String,
    pub span: Span,
}

impl From<(&str, Span)> for SourceSpan {
    fn from(value: (&str, Span)) -> Self {
        Self {
            source: value.0.parse().unwrap(),
            span: value.1,
        }
    }
}

impl ariadne::Span for SourceSpan {
    type SourceId = String;

    fn source(&self) -> &Self::SourceId {
        &self.source
    }

    fn start(&self) -> usize {
        self.span.start
    }

    fn end(&self) -> usize {
        self.span.end
    }
}
