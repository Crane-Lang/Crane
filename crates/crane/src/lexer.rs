mod error;

pub mod token;

pub use error::*;

use logos::Logos;

use crate::lexer::token::{Token, TokenKind};

pub struct Lexer<'src> {
    lexer: logos::Lexer<'src, TokenKind>,
}

impl<'src> Lexer<'src> {
    pub fn new(input: &'src str) -> Self {
        Self {
            lexer: TokenKind::lexer(input),
        }
    }
}

impl<'src> Iterator for Lexer<'src> {
    type Item = Result<Token, LexError>;

    fn next(&mut self) -> Option<Self::Item> {
        let kind = self.lexer.next()?;
        let lexeme = self.lexer.slice();
        let span = self.lexer.span().into();

        Some(
            kind.map(|kind| Token {
                kind,
                lexeme: lexeme.into(),
                span,
            })
            .map_err(|kind| LexError { kind, span }),
        )
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn lex(input: &str) -> logos::Lexer<TokenKind> {
        TokenKind::lexer(input)
    }

    fn check(input: &str, token: TokenKind) {
        let mut lexer = TokenKind::lexer(input);

        assert_eq!(lexer.next(), Some(Ok(token)));
        assert_eq!(lexer.slice(), input);
    }

    #[test]
    fn lex_open_paren() {
        check("(", TokenKind::OpenParen)
    }

    #[test]
    fn lex_close_paren() {
        check(")", TokenKind::CloseParen)
    }

    #[test]
    fn lex_open_brace() {
        check("{", TokenKind::OpenBrace)
    }

    #[test]
    fn lex_close_brace() {
        check("}", TokenKind::CloseBrace)
    }

    #[test]
    fn lex_comment() {
        check("// This is a comment.", TokenKind::Comment)
    }

    #[test]
    fn lex_fn_declaration() {
        let mut lex = lex("fn foo() {}");

        assert_eq!(lex.next(), Some(Ok(TokenKind::Ident)));
        assert_eq!(lex.span(), 0..2);
        assert_eq!(lex.slice(), "fn");

        assert_eq!(lex.next(), Some(Ok(TokenKind::Ident)));
        assert_eq!(lex.span(), 3..6);
        assert_eq!(lex.slice(), "foo");

        assert_eq!(lex.next(), Some(Ok(TokenKind::OpenParen)));
        assert_eq!(lex.span(), 6..7);
        assert_eq!(lex.slice(), "(");

        assert_eq!(lex.next(), Some(Ok(TokenKind::CloseParen)));
        assert_eq!(lex.span(), 7..8);
        assert_eq!(lex.slice(), ")");

        assert_eq!(lex.next(), Some(Ok(TokenKind::OpenBrace)));
        assert_eq!(lex.span(), 9..10);
        assert_eq!(lex.slice(), "{");

        assert_eq!(lex.next(), Some(Ok(TokenKind::CloseBrace)));
        assert_eq!(lex.span(), 10..11);
        assert_eq!(lex.slice(), "}");
    }

    #[test]
    fn test_lexer() {
        insta::glob!("snapshot_inputs/*.crane", |path| {
            let source = std::fs::read_to_string(path).unwrap();

            let lexer = Lexer::new(&source);

            insta::assert_yaml_snapshot!(lexer.into_iter().collect::<Vec<_>>());
        })
    }
}
