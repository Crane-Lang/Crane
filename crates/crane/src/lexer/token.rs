use logos::Logos;
use smol_str::SmolStr;

use crate::ast::Span;
use crate::lexer::LexErrorKind;

#[derive(Logos, Debug, PartialEq, Clone, Copy)]
#[logos(error = LexErrorKind)]
pub enum TokenKind {
    /// `(`
    #[token("(")]
    OpenParen,

    /// `)`
    #[token(")")]
    CloseParen,

    /// `{`
    #[token("{")]
    OpenBrace,

    /// `}`
    #[token("}")]
    CloseBrace,

    /// `,`
    #[token(",")]
    Comma,

    /// An identifier.
    #[regex(r"[A-Za-z_]+")]
    Ident,

    /// A string literal.
    #[regex(r#""[^"]*""#)]
    String,

    /// An integer literal.
    #[regex(r#"[\d]+"#)]
    Integer,

    /// A comment.
    #[regex(r"//.*")]
    Comment,

    /// Any sequence of whitespace characters.
    #[regex(r"[ \n]+", logos::skip)]
    Whitespace,
}

#[derive(Debug, Clone)]
pub struct Token {
    pub kind: TokenKind,
    pub lexeme: SmolStr,
    pub span: Span,
}
