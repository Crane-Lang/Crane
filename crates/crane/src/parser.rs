mod error;
mod expr;
mod item;
mod stmt;
mod ty;

pub use error::*;

use thin_vec::ThinVec;
use tracing::trace;

use crate::ast::{Ident, Item, Path, PathSegment, Span, DUMMY_SPAN};
use crate::lexer::token::{Token, TokenKind};
use crate::lexer::LexError;

pub type ParseResult<T> = Result<T, ParseError>;

/// An expected token.
#[derive(Debug, Clone, PartialEq)]
enum ExpectedToken {
    Token(TokenKind),
    Keyword(Ident),
    Ident,
}

impl std::fmt::Display for ExpectedToken {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{}",
            match self {
                Self::Token(kind) => format!("`{kind:?}`"),
                Self::Keyword(keyword) => format!("`{keyword}`"),
                Self::Ident => "an identifier".to_string(),
            }
        )
    }
}

pub struct Parser<TokenStream: Iterator<Item = Result<Token, LexError>>> {
    /// The list of tokens.
    tokens: TokenStream,

    /// The list of lexing errors uncovered during parsing.
    lex_errors: Vec<LexError>,

    /// The list of tokens the parser was expecting.
    expected_tokens: Vec<ExpectedToken>,

    /// The current token.
    token: Token,

    /// The previous token.
    prev_token: Token,
}

impl<TokenStream> Parser<TokenStream>
where
    TokenStream: Iterator<Item = Result<Token, LexError>>,
{
    pub fn new(input: TokenStream) -> Self {
        let mut parser = Self {
            tokens: input,
            lex_errors: Vec::new(),
            expected_tokens: Vec::new(),
            token: Token::dummy(),
            prev_token: Token::dummy(),
        };

        // Advance the parser to the first token.
        parser.advance();

        parser
    }

    pub fn parse(mut self) -> ParseResult<ThinVec<Item>> {
        trace!("Parsing program");

        let items_result = self.parse_module_items();

        let items = self.ensure_no_errors(items_result)?;

        if !self.is_at_end() {
            return Err(ParseError {
                kind: ParseErrorKind::Error("Failed to parse.".to_string()),
                span: self.token.span,
            });
        }

        Ok(items)
    }

    fn parse_module_items(&mut self) -> ParseResult<ThinVec<Item>> {
        let mut items = ThinVec::new();

        while let Some(item) = self.parse_item()? {
            items.push(item);
        }

        Ok(items)
    }

    fn ensure_no_errors<T>(&self, parse_result: ParseResult<T>) -> ParseResult<T> {
        if let Some(lex_error) = self.lex_errors.first() {
            return Err(ParseError {
                kind: ParseErrorKind::LexError(lex_error.kind.clone()),
                span: lex_error.span,
            });
        }

        parse_result
    }

    /// Returns whether the parser is at the end of the token stream.
    fn is_at_end(&mut self) -> bool {
        self.token.kind == TokenKind::Eof
    }

    /// Advances the parser to the next token.
    fn advance(&mut self) {
        let next_token: Option<Token>;

        loop {
            match self.tokens.next() {
                // Ignore any comment tokens.
                Some(Ok(Token {
                    kind: TokenKind::Comment | TokenKind::DocComment,
                    ..
                })) => {
                    continue;
                }

                // Capture any lexing errors, as we come across them.
                Some(Err(lex_error)) => {
                    next_token = None;

                    self.lex_errors.push(lex_error);

                    break;
                }
                Some(Ok(token)) => {
                    next_token = Some(token);

                    break;
                }
                None => {
                    next_token = None;

                    break;
                }
            };
        }

        // TODO: Figure out a better way of dealing with the end of input.
        let next_token = next_token.unwrap_or(Token {
            kind: TokenKind::Eof,
            lexeme: "".into(),
            span: Span {
                start: self.token.span.end,
                end: self.token.span.end + 1,
            },
        });

        self.prev_token = std::mem::replace(&mut self.token, next_token);
    }

    /// Returns whether the next token is of the given [`TokenKind`].
    ///
    /// If the token is not present this method will add the token to the list
    /// of expected tokens.
    fn check(&mut self, kind: TokenKind) -> bool {
        let is_present = self.token.kind == kind;
        if !is_present {
            self.expected_tokens.push(ExpectedToken::Token(kind));
        }

        is_present
    }

    /// Returns whether the next token is of the given [`TokenKind`].
    ///
    /// This does not add the token to the list of expected tokens.
    fn check_without_expect(&mut self, kind: TokenKind) -> bool {
        self.token.kind == kind
    }

    /// Returns whether the next token is the given keyword.
    ///
    /// This method adds the keyword to the list of expected tokens.
    fn check_keyword(&mut self, keyword: Ident) -> bool {
        self.expected_tokens
            .push(ExpectedToken::Keyword(keyword.clone()));

        self.token.is_keyword(keyword)
    }

    /// Consumes the next token if it is of the given [`TokenKind`].
    ///
    /// Returns whether the token was present.
    pub fn consume(&mut self, kind: TokenKind) -> bool {
        let is_present = self.check(kind);
        if is_present {
            self.advance();
        }

        is_present
    }

    /// Consumes the next token if it is the given keyword
    ///
    /// Otherwise it does not consume the token and returns `false`.
    ///
    /// This method adds the keyword to the list of expected tokens.
    pub fn consume_keyword(&mut self, keyword: Ident) -> bool {
        if self.check_keyword(keyword) {
            self.advance();

            true
        } else {
            false
        }
    }

    /// Parses an [`Ident`].
    pub fn parse_ident(&mut self) -> ParseResult<Ident> {
        let ident = self.token.ident().ok_or_else(|| ParseError {
            kind: ParseErrorKind::Error("Expected an identifier".to_string()),
            span: self.token.span,
        })?;

        self.advance();

        Ok(ident)
    }

    /// Parses a [`Path`].
    fn parse_path(&mut self) -> ParseResult<Path> {
        let mut path_segments = ThinVec::new();

        while let Some(ident) = self.parse_ident().ok() {
            path_segments.push(PathSegment { ident });

            if !self.consume(TokenKind::ColonColon) {
                break;
            }
        }

        let span = {
            let start_span = path_segments.first().map(|segment| segment.ident.span);
            let end_span = path_segments.last().map(|segment| segment.ident.span);

            match (start_span, end_span) {
                (Some(start), Some(end)) => start.to(end),
                (Some(span), None) | (None, Some(span)) => span,
                // TODO: Is this case impossible?
                (None, None) => DUMMY_SPAN,
            }
        };

        Ok(Path {
            segments: path_segments,
            span,
        })
    }
}

#[cfg(test)]
mod tests {
    use crate::lexer::Lexer;

    use super::*;

    #[test]
    pub fn test_parser() {
        insta::glob!("snapshot_inputs/*.crane", |path| {
            let source = std::fs::read_to_string(path).unwrap();

            let lexer = Lexer::new(&source);
            let parser = Parser::new(lexer);

            insta::assert_yaml_snapshot!(parser.parse());
        })
    }
}
