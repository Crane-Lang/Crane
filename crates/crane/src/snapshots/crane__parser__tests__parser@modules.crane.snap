---
source: crates/crane/src/parser.rs
expression: parser.parse()
input_file: crates/crane/src/snapshot_inputs/modules.crane
---
Ok:
  - kind:
      Use:
        prefix:
          segments:
            - ident:
                name: std
                span:
                  start: 4
                  end: 7
            - ident:
                name: io
                span:
                  start: 9
                  end: 11
            - ident:
                name: print
                span:
                  start: 13
                  end: 18
          span:
            start: 0
            end: 0
        kind: Single
    name:
      name: ""
      span:
        start: 0
        end: 0
  - kind:
      Use:
        prefix:
          segments:
            - ident:
                name: std
                span:
                  start: 23
                  end: 26
            - ident:
                name: io
                span:
                  start: 28
                  end: 30
            - ident:
                name: println
                span:
                  start: 32
                  end: 39
          span:
            start: 0
            end: 0
        kind: Single
    name:
      name: ""
      span:
        start: 0
        end: 0
  - kind:
      Use:
        prefix:
          segments:
            - ident:
                name: std
                span:
                  start: 44
                  end: 47
            - ident:
                name: int
                span:
                  start: 49
                  end: 52
            - ident:
                name: int_to_string
                span:
                  start: 54
                  end: 67
          span:
            start: 0
            end: 0
        kind: Single
    name:
      name: ""
      span:
        start: 0
        end: 0
  - kind:
      Fn:
        params: []
        return_ty: ~
        body:
          - kind:
              Expr:
                kind:
                  Call:
                    fun:
                      kind:
                        Variable:
                          segments:
                            - ident:
                                name: print
                                span:
                                  start: 85
                                  end: 90
                          span:
                            start: 85
                            end: 90
                      span:
                        start: 85
                        end: 90
                    args:
                      - kind:
                          Literal:
                            kind: String
                            value: "\"Your age is \""
                        span:
                          start: 91
                          end: 105
                span:
                  start: 85
                  end: 90
            span:
              start: 85
              end: 90
          - kind:
              Expr:
                kind:
                  Call:
                    fun:
                      kind:
                        Variable:
                          segments:
                            - ident:
                                name: println
                                span:
                                  start: 111
                                  end: 118
                          span:
                            start: 111
                            end: 118
                      span:
                        start: 111
                        end: 118
                    args:
                      - kind:
                          Call:
                            fun:
                              kind:
                                Variable:
                                  segments:
                                    - ident:
                                        name: int_to_string
                                        span:
                                          start: 119
                                          end: 132
                                  span:
                                    start: 119
                                    end: 132
                              span:
                                start: 119
                                end: 132
                            args:
                              - kind:
                                  Call:
                                    fun:
                                      kind:
                                        Variable:
                                          segments:
                                            - ident:
                                                name: user
                                                span:
                                                  start: 133
                                                  end: 137
                                            - ident:
                                                name: demographic
                                                span:
                                                  start: 139
                                                  end: 150
                                            - ident:
                                                name: get_age
                                                span:
                                                  start: 152
                                                  end: 159
                                          span:
                                            start: 133
                                            end: 159
                                      span:
                                        start: 133
                                        end: 159
                                    args: []
                                span:
                                  start: 133
                                  end: 159
                        span:
                          start: 119
                          end: 132
                span:
                  start: 111
                  end: 118
            span:
              start: 111
              end: 118
    name:
      name: main
      span:
        start: 72
        end: 76
  - kind:
      Module:
        Loaded:
          - items:
              - kind:
                  Module:
                    Loaded:
                      - items:
                          - kind:
                              Fn:
                                params: []
                                return_ty:
                                  name: Uint64
                                  span:
                                    start: 224
                                    end: 230
                                body:
                                  - kind:
                                      Expr:
                                        kind:
                                          Literal:
                                            kind: Integer
                                            value: "28"
                                        span:
                                          start: 245
                                          end: 247
                                    span:
                                      start: 245
                                      end: 247
                            name:
                              name: get_age
                              span:
                                start: 211
                                end: 218
                      - "Yes"
                name:
                  name: demographic
                  span:
                    start: 186
                    end: 197
          - "Yes"
    name:
      name: user
      span:
        start: 171
        end: 175

